builder/dockerfile/evaluator.go

```diff
@@ -52,10 +52,12 @@ func dispatch(d dispatchRequest, cmd instructions.Command) error {
 		}
 	}
 
-	if d.builder.options.ForceRemove {
-		defer d.builder.containerManager.RemoveAll(d.builder.Stdout)
-	}
-
+	defer func() {
+		if d.builder.options.ForceRemove || // unconditionally remove intermediate containers
+			(d.builder.options.Remove && err == nil) { // if no error while running command, remove intermediate containers
+			d.builder.containerManager.RemoveAll(d.builder.Stdout)
+		}
```

Comment:
Perhaps making these two separate conditions (small bit of duplication, but may be better to understand without having to add comments to explain)

```go
if d.builder.options.ForceRemove {
	d.builder.containerManager.RemoveAll(d.builder.Stdout)
	return
}
if d.builder.options.Remove && err == nil {
	d.builder.containerManager.RemoveAll(d.builder.Stdout)
	return
}
```

One thing I was wondering; it `--rm=false`, should `--force-remove` still remove containers? Can you check on an older version?

If that's the case, then it should likely look like:

```go
if !d.builder.options.Remove {
	return
}
if err == nil || d.builder.options.ForceRemove {
	d.builder.containerManager.RemoveAll(d.builder.Stdout)
}
```

Reply:
@thaJeztah on options deserialization, there is a reconcilation that forces `rm` to true if `force-rm` is true:
```
	if httputils.BoolValue(r, "forcerm") && versions.GreaterThanOrEqualTo(version, "1.12") {
		options.Remove = true
	} else if r.FormValue("rm") == "" && versions.GreaterThanOrEqualTo(version, "1.12") {
		options.Remove = true
	} else {
		options.Remove = httputils.BoolValue(r, "rm")
	}
```